modules/python/src2/cv2_convert.cpp

```diff
@@ -727,14 +727,30 @@ bool pyopencv_to(PyObject* obj, String &value, const ArgInfo& info)
 template<>
 PyObject* pyopencv_from(const String& value)
 {
-    return PyString_FromString(value.empty() ? "" : value.c_str());
+    const char* v = value.empty() ? "" : value.c_str();
+    PyObject* res = PyString_FromString(v);
+    if (!res)
+    {
+        PyErr_Clear();
+        return PyUnicode_DecodeRawUnicodeEscape(v, value.size(), NULL);
```

Comment:
OpenCV strings are UTF-8 (ASCII compatible).

We can't reinterpret them as there is no metadata about used other encodings.

---

Bug should be fixed in QR decoder instead (as it doesn't produce UTF-8 strings).

Reply:
Got it, thanks. Need one more round.

BTW, during the test design I tried use OpenCV’s encode and it could return the result properly.

Reply:
This is a bytes array which is returned after the decoding:

[83, 80, 67, 13, 10, 48, 50, 48, 48, 13, 10, 49, 13, 10, 67, 72, 48, 52, 51, 48, 48, 48, 53, 50, 51, 48, 50, 50, 50, 52, 52, 57, 48, 49, 72, 13, 10, 83, 13, 10, 69, 109, 105, 108, 32, 70, 114, 101, 121, 32, 66, 101, 116, 114, 105, 101, 98, 115, 32, 65, 71, 13, 10, 66, 97, 104, 110, 104, 111, 102, 115, 116, 114, 97, 115, 115, 101, 32, 49, 55, 13, 10, 13, 10, 53, 55, 52, 53, 13, 10, 83, 97, 102, 101, 110, 119, 105, 108, 13, 10, 67, 72, 13, 10, 13, 10, 13, 10, 13, 10, 13, 10, 13, 10, 13, 10, 13, 10, 51, 50, 50, 56, 46, 53, 48, 13, 10, 67, 72, 70, 13, 10, 83, 13, 10, 83, 105, 120, 116, 32, 114, 101, 110, 116, 32, 97, 32, 67, 97, 114, 32, 65, 71, 32, 13, 10, 77, -4, 108, 108, 104, 101, 105, 109, 115, 116, 114, 97, 115, 115, 101, 32, 49, 57, 53, 13, 10, 13, 10, 52, 48, 53, 55, 13, 10, 66, 97, 115, 101, 108, 13, 10, 67, 72, 13, 10, 81, 82, 82, 13, 10, 50, 54, 55, 50, 55, 52, 48, 51, 53, 56, 49, 48, 49, 48, 52, 56, 51, 48, 48, 48, 57, 54, 51, 57, 52, 51, 48, 13, 10, 75, 100, 110, 114, 32, 57, 54, 51, 57, 52, 51, 44, 32, 48, 51, 53, 56, 49, 45, 48, 49, 48, 52, 56, 51, 48, 13, 10, 69, 80, 68]

There is a `-4` value (252)

Reply:
Decoder returns bytes array.
We just should not call this as UTF-8 string (in all cases).

Reply:
@opencv-alalek, what do you think about such alternative solution: https://github.com/opencv/opencv/compare/4.x...dkurt:opencv:qrcode_eci_encoding?expand=1 ?

Key things:
1. `Py_BuildValue` returns nullptr if one of the values in the arguments is nullptr. So `pyopencv_from(String)` has to return `PyObject` or None, not nullptr:
```cpp
return Py_BuildValue("(NNN)", pyopencv_from(retval), pyopencv_from(points), pyopencv_from(straight_code));
```
2. Introduce `getEncoding` method to obtain saved ECI numbers and use it later in the bindings to properly decode bytes (need a test)
3. Shadow all `*decode*` (now only `detectAndDecode`) which may return 3 types of decoded info: original UTF-8, raw bytes if no decoding specified, encoded string if specified in QR code (where `getEncoding` wiil be used).

Reply:
It is not a Python problem at all.
Problem exists in C++ and should be fixes there.

In case of non UTF-8 bytes array which we want to retrieve as UTF-8 string, we should re-encode contents to UTF-8 string on C++ level internally (in decode API).

Reply:
> In case of non UTF-8 bytes array

We need to use this properly (under condition):

https://github.com/opencv/opencv/blob/1bccc14e05dc78908338656ee1c90ebae01357eb/3rdparty/quirc/include/quirc.h#L90-L111

Reply:
@opencv-alalek, you're right. Added a check for UTF-8 encoding and applied it only for `QUIRC_DATA_TYPE_BYTE` data.